// vant/es/signature/Signature.mjs
import { createVNode as _createVNode } from "vue";
import { ref, reactive, onMounted, nextTick, defineComponent } from "vue";
import { inBrowser, makeNumberProp, makeStringProp, createNamespace, preventDefault } from "../utils/index.mjs";
import { useRect } from "@vant/use";
import { Button } from "../button/index.mjs";
const [name, bem, t] = createNamespace("signature");
const signatureProps = {
  tips: String,
  type: makeStringProp("png"),
  penColor: makeStringProp("#000"),
  lineWidth: makeNumberProp(3),
  clearButtonText: String,
  backgroundColor: makeStringProp(""),
  confirmButtonText: String
};
const hasCanvasSupport = () => {
  var _a;
  const canvas = document.createElement("canvas");
  return !!((_a = canvas.getContext) == null ? void 0 : _a.call(canvas, "2d"));
};
var stdin_default = defineComponent({
  name,
  props: signatureProps,
  emits: ["submit", "clear", "start", "end", "signing"],
  setup(props, {
    emit
  }) {
    const canvasRef = ref();
    const wrapRef = ref();
    const state = reactive({
      width: 0,
      height: 0,
      ctx: null,
      ratio: inBrowser ? window.devicePixelRatio : 1
    });
    let canvasRect;
    const isRenderCanvas = inBrowser ? hasCanvasSupport() : true;
    const touchStart = () => {
      if (!state.ctx) {
        return false;
      }
      state.ctx.beginPath();
      state.ctx.lineWidth = props.lineWidth * state.ratio;
      state.ctx.strokeStyle = props.penColor;
      canvasRect = useRect(canvasRef);
      emit("start");
    };
    const touchMove = (event) => {
      var _a, _b;
      if (!state.ctx) {
        return false;
      }
      preventDefault(event);
      const touch = event.touches[0];
      const mouseX = (touch.clientX - ((canvasRect == null ? void 0 : canvasRect.left) || 0)) * state.ratio;
      const mouseY = (touch.clientY - ((canvasRect == null ? void 0 : canvasRect.top) || 0)) * state.ratio;
      state.ctx.lineCap = "round";
      state.ctx.lineJoin = "round";
      (_a = state.ctx) == null ? void 0 : _a.lineTo(mouseX, mouseY);
      (_b = state.ctx) == null ? void 0 : _b.stroke();
      emit("signing", event);
    };
    const touchEnd = (event) => {
      preventDefault(event);
      emit("end");
    };
    const isCanvasEmpty = (canvas) => {
      const empty = document.createElement("canvas");
      empty.width = canvas.width;
      empty.height = canvas.height;
      return canvas.toDataURL() === empty.toDataURL();
    };
    const setCanvasBgColor = () => {
      if (state.ctx && props.backgroundColor) {
        state.ctx.fillStyle = props.backgroundColor;
        state.ctx.fillRect(0, 0, state.width, state.height);
      }
    };
    const submit = () => {
      var _a, _b;
      const canvas = canvasRef.value;
      if (!canvas) {
        return;
      }
      const isEmpty = isCanvasEmpty(canvas);
      const image = isEmpty ? "" : ((_b = (_a = {
        jpg: () => canvas.toDataURL("image/jpeg", 0.8),
        jpeg: () => canvas.toDataURL("image/jpeg", 0.8)
      })[props.type]) == null ? void 0 : _b.call(_a)) || canvas.toDataURL(`image/${props.type}`);
      emit("submit", {
        image,
        canvas
      });
    };
    const clear = () => {
      if (state.ctx) {
        state.ctx.clearRect(0, 0, state.width, state.height);
        state.ctx.closePath();
        setCanvasBgColor();
      }
      emit("clear");
    };
    onMounted(() => {
      var _a, _b, _c;
      if (isRenderCanvas) {
        state.ctx = (_a = canvasRef.value) == null ? void 0 : _a.getContext("2d");
        state.width = (((_b = wrapRef.value) == null ? void 0 : _b.offsetWidth) || 0) * state.ratio;
        state.height = (((_c = wrapRef.value) == null ? void 0 : _c.offsetHeight) || 0) * state.ratio;
        nextTick(() => {
          setCanvasBgColor();
        });
      }
    });
    return () => _createVNode("div", {
      "class": bem()
    }, [_createVNode("div", {
      "class": bem("content"),
      "ref": wrapRef
    }, [isRenderCanvas ? _createVNode("canvas", {
      "ref": canvasRef,
      "width": state.width,
      "height": state.height,
      "onTouchstartPassive": touchStart,
      "onTouchmove": touchMove,
      "onTouchend": touchEnd
    }, null) : _createVNode("p", null, [props.tips])]), _createVNode("div", {
      "class": bem("footer")
    }, [_createVNode(Button, {
      "size": "small",
      "onClick": clear
    }, {
      default: () => [props.clearButtonText || t("clear")]
    }), _createVNode(Button, {
      "type": "primary",
      "size": "small",
      "onClick": submit
    }, {
      default: () => [props.confirmButtonText || t("confirm")]
    })])]);
  }
});
export {
  stdin_default as default,
  signatureProps
};
