// vant/lib/floating-panel/FloatingPanel.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name2 in all)
    __defProp(target, name2, { get: all[name2], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var stdin_exports = {};
__export(stdin_exports, {
  default: () => stdin_default,
  floatingPanelProps: () => floatingPanelProps
});
module.exports = __toCommonJS(stdin_exports);
var import_vue = require("vue");
var import_vue2 = require("vue");
var import_utils = require("../utils");
var import_use = require("@vant/use");
var import_use_lock_scroll = require("../composables/use-lock-scroll");
var import_use_touch = require("../composables/use-touch");
var import_use_sync_prop_ref = require("../composables/use-sync-prop-ref");
const {
  height: windowHeight
} = (0, import_use.useWindowSize)();
const floatingPanelProps = {
  height: (0, import_utils.makeNumericProp)(0),
  anchors: (0, import_utils.makeArrayProp)(),
  duration: (0, import_utils.makeNumericProp)(0.2),
  contentDraggable: import_utils.truthProp,
  safeAreaInsetBottom: import_utils.truthProp
};
const [name, bem] = (0, import_utils.createNamespace)("floating-panel");
const DAMP = 0.2;
var stdin_default = (0, import_vue2.defineComponent)({
  name,
  props: floatingPanelProps,
  emits: ["heightChange", "update:height"],
  setup(props, {
    emit,
    slots
  }) {
    const rootRef = (0, import_vue2.ref)();
    const contentRef = (0, import_vue2.ref)();
    const height = (0, import_use_sync_prop_ref.useSyncPropRef)(() => +props.height, (value) => emit("update:height", value));
    const boundary = (0, import_vue2.computed)(() => {
      var _a, _b;
      return {
        min: (_a = props.anchors[0]) != null ? _a : 100,
        max: (_b = props.anchors[props.anchors.length - 1]) != null ? _b : Math.round(windowHeight.value * 0.6)
      };
    });
    const anchors = (0, import_vue2.computed)(() => props.anchors.length >= 2 ? props.anchors : [boundary.value.min, boundary.value.max]);
    const dragging = (0, import_vue2.ref)(false);
    const rootStyle = (0, import_vue2.computed)(() => ({
      height: (0, import_utils.addUnit)(boundary.value.max),
      transform: `translateY(calc(100% + ${(0, import_utils.addUnit)(-height.value)}))`,
      transition: !dragging.value ? `transform ${props.duration}s` : "none"
    }));
    const ease = (moveY) => {
      const absDistance = Math.abs(moveY);
      const {
        min,
        max
      } = boundary.value;
      if (absDistance > max) {
        return -(max + (absDistance - max) * DAMP);
      }
      if (absDistance < min) {
        return -(min - (min - absDistance) * DAMP);
      }
      return moveY;
    };
    const closest = (arr, target) => arr.reduce((pre, cur) => Math.abs(pre - target) < Math.abs(cur - target) ? pre : cur);
    let startY;
    const touch = (0, import_use_touch.useTouch)();
    const onTouchstart = (e) => {
      touch.start(e);
      dragging.value = true;
      startY = -height.value;
    };
    const onTouchmove = (e) => {
      var _a;
      touch.move(e);
      const target = e.target;
      if (contentRef.value === target || ((_a = contentRef.value) == null ? void 0 : _a.contains(target))) {
        if (!props.contentDraggable)
          return;
        if (-startY < boundary.value.max) {
          if (e.cancelable)
            e.preventDefault();
          e.stopPropagation();
        } else if (!(contentRef.value.scrollTop <= 0 && touch.deltaY.value > 0)) {
          return;
        }
      }
      const moveY = touch.deltaY.value + startY;
      height.value = -ease(moveY);
    };
    const onTouchend = () => {
      dragging.value = false;
      height.value = closest(anchors.value, height.value);
      if (height.value !== -startY) {
        emit("heightChange", {
          height: height.value
        });
      }
    };
    (0, import_vue2.watch)(boundary, () => {
      height.value = closest(anchors.value, height.value);
    }, {
      immediate: true
    });
    (0, import_use_lock_scroll.useLockScroll)(rootRef, () => true);
    return () => {
      var _a;
      return (0, import_vue.createVNode)("div", {
        "class": [bem(), {
          "van-safe-area-bottom": props.safeAreaInsetBottom
        }],
        "ref": rootRef,
        "style": rootStyle.value,
        "onTouchstartPassive": onTouchstart,
        "onTouchmove": onTouchmove,
        "onTouchend": onTouchend,
        "onTouchcancel": onTouchend
      }, [(0, import_vue.createVNode)("div", {
        "class": bem("header")
      }, [(0, import_vue.createVNode)("div", {
        "class": bem("header-bar")
      }, null)]), (0, import_vue.createVNode)("div", {
        "class": bem("content"),
        "ref": contentRef
      }, [(_a = slots.default) == null ? void 0 : _a.call(slots)])]);
    };
  }
});
